crates/compiler/builtins/roc/Task.roc

```diff
@@ -0,0 +1,251 @@
+## program, such as the terminal's standard output, or a file.
+Task ok err := {} -> Result ok err
+
+## Run a task repeatedly, until it fails with `err`. Note that this task does not return a success value.
+forever : Task a err -> Task * err
+forever = \@Task task ->
+    looper = \{} ->
+        when task {} is
+            Err e -> Err e
+            Ok _ -> looper {}
+
+    @Task \{} -> looper {}
+
+# TODO: consider moving referenced documentation here.
+#
+## Run a task repeatedly, until it fails with `err` or completes with `done`.
+## Check out [this example](https://www.roc-lang.org/examples/TaskLoop/README.html).
+loop : state, (state -> Task [Step state, Done done] err) -> Task done err
+loop = \state, step ->
+    looper = \current ->
+        (@Task next) = step current
+        when next {} is
+            Err e -> Err e
+            Ok (Done newResult) -> Ok newResult
+            Ok (Step newState) -> looper (newState)
+
+    @Task \{} -> looper state
+
+## Create a task that always succeeds with the value provided.
+##
+## ```
+## # Always succeeds with "Louis"
+## getName : Task.Task Str *
+## getName = Task.ok "Louis"
+## ```
+##
+ok : a -> Task a *
+ok = \a -> @Task \{} -> Ok a
+
+## Create a task that always fails with the error provided.
+##
+## ```
+## # Always fails with the tag `CustomError Str`
+## customError : Str -> Task.Task {} [CustomError Str]
+## customError = \err -> Task.err (CustomError err)
+## ```
+##
+err : a -> Task * a
+err = \a -> @Task \{} -> Err a
+
+## Transform a given Task with a function that handles the success or error case
+## and returns another task based on that. This is useful for chaining tasks
+## together or performing error handling and recovery.
+##
+## Consider a the following task;
+##
+## `canFail : Task {} [Failure, AnotherFail, YetAnotherFail]`
+##
+## We can use [attempt] to handle the failure cases using the following;
+##
+## ```
+## result <- canFail |> Task.attempt
+## when result is
+##     Ok Success -> Stdout.line "Success!"
+##     Err Failure -> Stdout.line "Oops, failed!"
+##     Err AnotherFail -> Stdout.line "Ooooops, another failure!"
+##     Err YetAnotherFail -> Stdout.line "Really big oooooops, yet again!"
+## ```
+##
+## Here we know that the `canFail` task may fail, and so we use
+## `Task.attempt` to convert the task to a `Result` and then use pattern
+## matching to handle the success and possible failure cases.
+##
+attempt : Task a b, (Result a b -> Task c d) -> Task c d
+attempt = \@Task task, transform ->
+    @Task \{} ->
+        (@Task transformed) = transform (task {})
+
+        transformed {}
+
+## Take the success value from a given [Task] and use that to generate a new [Task].
+##
+## For example we can use this to run tasks in sequence like follows;
+##
+## ```
+## # Prints "Hello World!\n" to standard output.
+## {} <- Stdout.write "Hello "|> Task.await
+## {} <- Stdout.write "World!\n"|> Task.await
+##
+## Task.ok {}
+## ```
```

Comment:
This example could be a bit better, for example to echo stdin to stdout. The current example doesn't use a material success value from the first task.

Reply:
Good idea. I both made that the example, and mentioned how ! desugars to this method.